src/main/java/com/mixpanel/android/mpmetrics/MixpanelNotificationRouteActivity.java

```diff
@@ -0,0 +1,163 @@
+        }
+
+        CharSequence uri = routeIntent.getExtras().getCharSequence("uri");
+
+        final Intent defaultIntent = this.getPackageManager().getLaunchIntentForPackage(this.getPackageName());
+
+        switch (target) {
+            case HOMESCREEN:
+                return defaultIntent;
+            case URL_IN_BROWSER:
+                if (isValidURL(uri.toString(), PushTapTarget.URL_IN_BROWSER)) {
+                    return new Intent(Intent.ACTION_VIEW, Uri.parse(uri.toString()));
+                } else {
+                    MPLog.d(LOGTAG, "Wanted to open url in browser but url is invalid: " + uri.toString() + ". Starting default intent");
+                    return defaultIntent;
+                }
+            case URL_IN_WEBVIEW:
+                if (isValidURL(uri.toString(), PushTapTarget.URL_IN_WEBVIEW)) {
+                    return new Intent(this.getApplicationContext(), MixpanelWebViewActivity.class).
+                            putExtra("uri", uri.toString()).
+                            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+                } else {
+                    MPLog.d(LOGTAG, "Wanted to open url in webview but url is invalid or not secure (http): " + uri.toString() + ". Starting default intent");
+                    return defaultIntent;
+                }
+            case DEEP_LINK:
+                return new Intent(Intent.ACTION_VIEW, Uri.parse(uri.toString()));
+            default:
+                return defaultIntent;
+        }
+    }
+
+    protected boolean isValidURL(CharSequence url, PushTapTarget target) {
+        if (target == PushTapTarget.URL_IN_WEBVIEW) {
+            return null != url && url.toString().startsWith("https");
+        } else {
+            return null != url && url.toString().startsWith("http");
+        }
+    }
+
+    protected void cancelNotification(Bundle extras) {
+        int notificationId = extras.getInt("notificationId");
+        NotificationManager notificationManager = (NotificationManager) getApplicationContext().getSystemService(getApplicationContext().NOTIFICATION_SERVICE);
+        notificationManager.cancel(notificationId);
+    }
+
+    protected void trackAction(Intent routeIntent) {
+        Bundle intentExtras = routeIntent.getExtras();
+
+        CharSequence tapTargetChars = intentExtras.getCharSequence("tapTarget");
+        if (null == tapTargetChars) {
+            MPLog.d(LOGTAG, "Notification action click logged with no tapTarget");
+            return;
+        }
+
+        final String tapTarget = tapTargetChars.toString();
+        final String buttonId;
+        final String label;
+
+        if (tapTarget.equals(MixpanelPushNotification.TAP_TARGET_BUTTON)) {
+            CharSequence buttonIdChars = intentExtras.getCharSequence("buttonId");
+            if (null == buttonIdChars) {
+                MPLog.d(LOGTAG, "Notification action click logged with no buttonId");
+                return;
```

Comment:
I would say in such a case, let's allow the track to continue, just set `buttonId` (and `label` below) to null